Cluster Validation Scan

Akatuoro opened the pull request:
<!--   Thank you for your contribution. Before you submit the pull request:
1. Follow contributing guidelines, templates, the recommended Git workflow, and any related documentation.
2. Read and submit the required Contributor Licence Agreements (https://github.com/kyma-project/community/blob/main/docs/contributing/02-contributing.md).
3. Test your changes and attach their results to the pull request.
4. Update the relevant documentation.
-->

**Description**

Changes proposed in this pull request:

- Add a cluster validation section in the cluster overview page
  - It is hidden behind a `CLUSTER_VALIDATION` feature flag
- The section enables a scan of the cluster
  - The scan validates the cluster resources against the rules of #2435
  - The user can configure which namespaces to scan and which policies to check
    - System namespaces like kube-system and kyma-system are excluded by default.
    - By default, the policies chosen by the user and cluster for resource validation are used
  - The user can configure the number of parallel requests to the api, 5 by default
- The scan result is shown as a tree where the user can drill down to the specific warnings
- This feature uses ui5 react webcomponents, so it includes the ui5 react webcomponent libraries and sets the necessary ThemeProvider
- The pqueue library is included for queuing requests and work for the worker thread

**Related issue(s)**

<!-- If you refer to a particular issue, provide its number. For example, `Resolves #123`, `Fixes #43`, or `See also #33`. -->
Resolves #2376

mrCherry97 reviewed (changes requested):
Scan Progress shouldn't be displayed if a user is not scanning cluster.

When I provide `Parallel Requests` <= 0 then UI breaks.

We have a talk to add button for choose all namespaces and uncheck all.

It will be nice to have some margin/padding here:
![image](https://github.com/kyma-project/busola/assets/46759954/15bcc32d-f4f3-4ef7-9e5b-e03f55f293bf)

Info about feature flag should be added in docs and you can add it to defaultConfig with enabled false for now.

Why you are implementing TypeScript and JavaScript mixed? `ClusterValidation.js` rest of new files are `ts`.

Akatuoro commented:
> * [x]  Scan Progress shouldn't be displayed if a user is not scanning cluster.

Adjusted so that both Scan Progress and Scan Result is not shown if not available.

> * [x]  When I provide `Parallel Requests` <= 0 then UI breaks.

Fixed

> * [ ]  We have a talk to add button for choose all namespaces and uncheck all.

will work on that tomorrow

> * [x]  It will be nice to have some margin/padding here:
>   ![image](https://user-images.githubusercontent.com/46759954/256514988-15bcc32d-f4f3-4ef7-9e5b-e03f55f293bf.png)

done

> * [x]  Info about feature flag should be added in docs and you can add it to defaultConfig with enabled false for now.

done

> * [x]  Why you are implementing TypeScript and JavaScript mixed? `ClusterValidation.js` rest of new files are `ts`.

Good point, I've changed `ClusterValidation.js` to `tsx` and `ResourceValidation.worker.js` to `ts`.
I had it in `js` for faster iteration with components that are not well typed. Also needed to update the fundamental-react types, but now it's working well.

There is still 1 file in `js`:
`ClusterValidationConfiguration.js` - this is explicitly in `js` because the `ResourceForm` does not have proper types. So it is very verbose to assign all `ResourceForm` properties in `ts`.

mrCherry97 reviewed (changes requested):
When I am a user with reduced permissions only for one namespace I can't list namespaces, but ClusterValidation feature is displaying all of them.
![image](https://github.com/kyma-project/busola/assets/46759954/0c7e802d-ebd4-448d-9596-6dad192b0f39)

![image](https://github.com/kyma-project/busola/assets/46759954/7501db2a-d629-4b3b-9a2f-2166f9ad19ec)

User should be able to see only this namespace where he has access.

mrCherry97 commented:
And also parallel <0 has no sense

Akatuoro commented:
> When I am a user with reduced permissions only for one namespace I can't list namespaces, but ClusterValidation feature is displaying all of them. ![image](https://user-images.githubusercontent.com/46759954/256803096-0c7e802d-ebd4-448d-9596-6dad192b0f39.png)
> 
> User should be able to see only this namespace where he has access.

Explanation: The restricted user has no access to /api/v1/namespaces , so the response status code is 403. This is counted as an error, so it goes into [processError](https://github.com/kyma-project/busola/blob/main/src/shared/hooks/BackendAPI/useGet.js#L77-L82) instead of [processDataFn](https://github.com/kyma-project/busola/blob/main/src/shared/hooks/BackendAPI/useGet.js#L74-L76)

I've included a fix for `useAvailableNamespaces`.

Akatuoro commented:
> And also parallel <0 has no sense

I've added a resource form validation for this. It is still possible to submit with <= 0, in which case the scan uses 1 connection.
![Screenshot 2023-07-28 at 15 53 55](https://github.com/kyma-project/busola/assets/4891926/cb2eabab-2f13-491f-8ef7-162bec1806f4)

mrCherry97 reviewed (changes requested):
We need to have some tests for that functionality.

Akatuoro commented:
> We need to have some tests for that functionality.

I've added tests for the scan

mrCherry97 commented:
/test all

Akatuoro commented:
/retest

mrCherry97 commented:
/ok-to-test

mrCherry97 commented:
/test all

mrCherry97 commented:
/retest

mrCherry97 commented:
/test all

mrCherry97 commented:
/retest

mrCherry97 commented:
/retest

Akatuoro commented:
/retest

akucharska commented on src/components/Clusters/views/ClusterOverview/ClusterValidation/ClusterValidation.tsx:
```diff
@@ -0,0 +1,292 @@
+
+      <LayoutPanel.Body>
+        <Section titleText={t('cluster-validation.scan.scope')}>
+          <FlexBox>
```
Consider adding padding to the flexbox
```suggestion
                    <FlexBox className="fd-has-padding-small">
```

akucharska reviewed (commented):
It looks really good.
My only suggestion is adding some additional styling for better looks

mrCherry97 commented:
/test all

mrCherry97 commented:
/test pull-busola-integration-namespace-k3d

mrCherry97 commented:
/retest

IwonaLanger commented on docs/resource-validation/README.md:
```diff
@@ -55,3 +55,9 @@ policies:
+
+## Scan the cluster
+
+With the `CLUSTER_VALIDATION` [feature flag](../features.md) enabled, you can use these rules to scan existing resources in your cluster.
```
I'd use the same formatting for CLUSTER_VALIDATION as in `features.md`, that is, bold font. It's not a flag, right?

mrCherry97 commented:
/retest
